server/team_matchmaking_service.py

```diff
@@ -0,0 +1,117 @@
+from typing import NamedTuple
+
+from server import GameService
+from server.decorators import with_logger
+from server.lobbyconnection import ClientError
+from server.players import Player
+from server.team_matchmaker.player_party import PlayerParty
+
+MapDescription = NamedTuple('Map', [("id", int), ("name", str), ("path", str)])
+GroupInvite = NamedTuple('GroupInvite', [("inviting", Player), ("invited", Player), ("party", PlayerParty)])
+
+
+@with_logger
+class TeamMatchmakingService:
+    """
+    Service responsible for managing the global team matchmaking. Does grouping, matchmaking, updates statistics, and
+    launches the games.
+    """
+
+    def __init__(self, games_service: GameService):
+        self.game_service = games_service
+        self.player_parties = dict() # player -> current party
+        self._pending_invites = dict() # invited player -> current pending invite to this player
+
+    def invite_player_to_group(self, inviting_player: Player, invited_player: Player):
+        if inviting_player not in self.player_parties:
+            self.player_parties[inviting_player] = PlayerParty(self, inviting_player)
+
+        inviting_party = self.player_parties.get(inviting_player)
+
+        if not inviting_party.owner == inviting_player:
+            raise ClientError("You do not own this party.", recoverable=True)
+
+        if inviting_player not in invited_player.friends:
+            raise ClientError("This person hasn't befriended you.", recoverable=True)
+
+        self.clear_invites_to(invited_player)
```

Comment:
As I guessed above; this means you can clear pending invites and theoretically DoS somebody out of party invites. Would it not be better to have a set of pending invites per player, which are cleared on acceptance rather than on creation of a new invite?

Reply:
Yes, I thought about that aswell, if you allow multiple invites you can still dos a user out of a party by flooding them with notifications. They will foe you this way or the other and then you're safe.

I'm also now considering removing the requirement of being befriended by a person to invite them and replace it with not being foed

Reply:
You just want a uniqueness constraint on behalf of the sender per target -- that is, you cannot invite a user multiple times to the same party?